httprange/httprange.go

```diff
@@ -0,0 +1,159 @@
+// Package httprange provides RFC7233 range parsing.
+package httprange
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+)
+
+// Bytes is the bytes range type specified in the RFC.
+type Bytes struct {
+	Start     int64
+	End       int64
+	Length    int64
+	Satisfied bool
+}
+
+func (b Bytes) fmtRequest() (string, error) {
+	switch {
+	case b.Start < 0 && b.End == -1:
+		return fmt.Sprintf("%d", b.Start), nil
+	case b.Start >= 0 && b.End == -1:
+		return fmt.Sprintf("%d-", b.Start), nil
+	case b.Start >= 0 && b.End >= 0:
+		if b.End < b.Start {
+			break
+		}
+		return fmt.Sprintf("%d-%d", b.Start, b.End), nil
+	default:
+	}
+	return "", fmt.Errorf("invalid request range")
+}
+
+func (b Bytes) fmtResponse() (string, error) {
+	if !b.Satisfied {
+		if b.Length < 0 {
+			return "", fmt.Errorf("invalid response length")
+		}
+		return fmt.Sprintf("bytes */%d", b.Length), nil
+	}
+	if b.Start < 0 || b.End < 0 {
+		return "", fmt.Errorf("invalid response range")
+	}
+	if b.Length < 0 {
+		return fmt.Sprintf("bytes %d-%d/*", b.Start, b.End), nil
+	}
+	return fmt.Sprintf("bytes %d-%d/%d", b.Start, b.End, b.Length), nil
+}
+
+// Range is the generic range type
+type Range struct {
+	Unit string
+	Resp string
+}
+
+// ErrNotByteUnit is returned from ParseBytes if the returned unit type is not "bytes".
+var ErrNotByteUnit = fmt.Errorf(`httprange: expected "bytes" unit type`)
+
+// ParseBytes parses a Content-Range header, expecting it to be a bytes range.
+func ParseBytes(h string) (Bytes, error) {
+	l := lex(h)
+	r := Bytes{}
+	for {
+		switch t := l.step(); t.kind {
+		case itemUnit:
+			if t.tok != "bytes" {
+				return Bytes{}, ErrNotByteUnit
+			}
+		case itemStart:
+			if t.tok == "*" {
+				r.Start = -1
+				r.End = -1
+				break
+			}
+			r.Satisfied = true
+			i, err := strconv.ParseInt(t.tok, 10, 64)
+			if err != nil {
+				return Bytes{}, err
+			}
+			r.Start = i
+		case itemEnd:
+			i, err := strconv.ParseInt(t.tok, 10, 64)
+			if err != nil {
+				return Bytes{}, err
+			}
+			r.End = i
+		case itemLength:
+			if t.tok == "*" {
+				r.Length = -1
+				break
+			}
+			i, err := strconv.ParseInt(t.tok, 10, 64)
+			if err != nil {
+				return Bytes{}, err
+			}
+			r.Length = i
+		case itemEOF:
+			return r, nil
+		case itemError:
+			return Bytes{}, fmt.Errorf("%s", t.tok)
+		default:
+			return Bytes{}, fmt.Errorf("lexer error: what's a %q?", t)
+		}
+	}
+}
+
+// FormatBytesRequest constructs a string suitable for using as a Range header.
+func FormatBytesRequest(r ...Bytes) (string, error) {
```

Comment:
NIT: Is there a particular design reason for using a variadic function over a slice?